crowbar_framework/app/models/service_object.rb

```diff
@@ -990,21 +1019,17 @@ def apply_role(role, inst, in_queue, bootstrap = false)
       Rails.logger.debug "delay empty - running proposal"
     end
 
-    # expand items in elements that are not nodes
-    expanded_new_elements = {}
-    new_deployment["elements"].each do |role_name, nodes|
-      expanded_new_elements[role_name], failures = expand_nodes_for_all(nodes)
-      unless failures.nil? || failures.empty?
-        Rails.logger.fatal(
-          "apply_role: Failed to expand items #{failures.inspect} for role \"#{role_name}\""
-        )
-        message = "Failed to apply the proposal: cannot expand list of nodes for role \"#{role_name}\", following items do not exist: #{failures.join(", ")}"
-        Rails.logger.progress("Failed to apply role #{role.name}")
-        update_proposal_status(inst, "failed", message)
-        return [405, message]
-      end
+    new_elements, failures = expand_items_in_elements(new_deployment["elements"])
+    unless failures.nil?
+      Rails.logger.fatal(
+        "apply_role: Failed to expand items #{failures.inspect} for role \"#{role_name}\""
+      )
```

Comment:
Argh. When we moved the error message here, outside of the helper, we lost access to `role_name`. I'd suggest to simply remove the mention? Or to go back and build the error message in the helper.

You could actually have the "cannot expand list of nodes..." built in the helper, and then add the prefix here (since we have two slightly different messages).

Reply:
we still need the message to set `update_proposal_status` with a proper error message. So Im guessing moving the logging into the helper, then returning the msg like before should work for this case